### ISSUE TITLE
Problem with memoize and keyword arguments

### ISSUE THREAD

**Issue by wlupton** [grantjenks/python-diskcache#195]
I'm using memoize with keyword arguments and am having problems with the `ENOVAL` part of the key when using `for key in cache` to iterate over the keys.

This is the original key (`ENOVAL` is a `Constant`):

<img width="486" alt="Screenshot 2021-02-25 at 12 20 14" src="https://user-images.githubusercontent.com/8874729/109152853-f3c25900-7763-11eb-97ee-0bd5d6f2ec2f.png">

But when I subsequently iterate over the keys I get this (`ENOVAL` is now a tuple with a string member):

<img width="488" alt="Screenshot 2021-02-25 at 12 20 57" src="https://user-images.githubusercontent.com/8874729/109152910-089eec80-7764-11eb-9c45-1a316ba4b8a7.png">

This is an invalid key (`cache.get(key)` returns nothing).

I experimented with changing `args_to_key()` as shown, and the problem goes away.

```
    if kwargs:
        # XXX this seems to become a nested tuple? use string for now
        # key += (ENOVAL,)
        key += ('ENOVAL',)
```

**Comment by grantjenks** [comment]
Are you using JSONDisk? I think this is how it serializes ENOVAL.

**Comment by wlupton** [comment]
I'm using the default, whatever that is.

**Comment by wlupton** [comment]
To be clear, without the fix, this code fails to retrieve values (it returns `None` for keys that should exist and have non-expired values).

```
with Cache(cache_directory) as cache:
    for key in cache:
        value, expire = cache.get(key, expire_time=True)
```

Would a simple demo program be helpful?

**Comment by grantjenks** [comment]
Of course :)

**Comment by grantjenks** [comment]
Current hypothesis: seems like the ENOVAL does not round-trip properly when serialized and deserialized.

**Comment by wlupton** [comment]
The program quoted below illustrates the problem:

```
% ./diskcache-enoval-serialize-bug.py 
('__main__.twice', 1) -> 2
('__main__.thrice', ENOVAL, 'arg', 1) -> None (KeyError)
```

With the fix mentioned earlier:

```
% ./diskcache-enoval-serialize-bug.py 
('__main__.twice', 1) -> 2
('__main__.thrice', 'ENOVAL', 'arg', 1) -> 3
```

Program:

```
#!/usr/bin/env python

from diskcache import Cache

cache = Cache()

@cache.memoize(expire=3600)
def twice(arg: int) -> int:
    return 2 * arg

@cache.memoize(expire=3600)
def thrice(*, arg: int = 1) -> int:
    return 3 * arg

twice(1)
thrice(arg=1)

for key in cache:
    error = ''
    try:
        cache[key]
    except KeyError:
        error = ' (KeyError)'
    print('%r -> %r%s' % (key, cache.get(key), error))
```

**Comment by xvr-hlt** [comment]
I'm having the same issue – my usage of memoize results in the following behaviour 

```
cache = ...

for k in cache.iterkeys():
  assert cache[k] is None
```

However, if I try and reassemble the key from the function using `__cache_key__` I can get the expected behaviour

```
for k in cache.iterkeys():
  k = k[2:] # remove function name and ENOENT from key to just get kwargs
  kwargs = dict(zip(k[::2], k[1::2]) # kwargs is now a dict formed from the original ke
  new_key = foo.__cache_key__(**kwargs)
  assert cache[new_key] is not None
```

**Comment by wlupton** [comment]
Could I suggest that maybe this fix (quoted earlier) might be applied? @xvr-hlt, does this fix address your problem?

Change `args_to_key()` as shown:

```
    if kwargs:
        # XXX this seems to become a nested tuple? use string for now
        # key += (ENOVAL,)
        key += ('ENOVAL',)
```

**Comment by grantjenks** [comment]
Fixed by d55a50e. Will be deployed to PyPI this month.
